solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/SolaceAsyncMessageProducer.java

```diff
@@ -0,0 +1,126 @@
+package com.solace.spring.cloud.stream.binder.util;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Consumer;
+
+import com.solace.spring.cloud.stream.binder.messaging.CorrelationData;
+import com.solacesystems.jcsmp.Destination;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPProperties;
+import com.solacesystems.jcsmp.JCSMPSession;
+import com.solacesystems.jcsmp.JCSMPStreamingPublishCorrelatingEventHandler;
+import com.solacesystems.jcsmp.XMLMessage;
+import com.solacesystems.jcsmp.XMLMessageProducer;
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+public class SolaceAsyncMessageProducer {
```

Comment:
Was the purpose of this class to:
* Have a map of message IDs to messages so that you can retrieve the messages in the publisher's event callback?
* Have a scheduler to timeout messages and also to manage the map of messages?

If it is just these, then I don't think you need this class, because:
* The message itself (as well as anything else you want) can be directly provided as part of the message's correlation key. All properties of the message can be self-contained to the message without the need of creating any external collections to keep track of them.
* The message timeout is already handled by having a a `Future` in the `CorrelationData` object. It can be configured and set by the publisher's message handler using `Future.get(timeout)`. In my opinion, having a binding-scoped configuration in this scenario (when ) would only conflict with this.

For example, here's a PR I opened [that addresses the issue of async publisher errors](https://github.com/SolaceProducts/solace-spring-cloud/pull/63). The main difference here is that everything needed to handle acknowledgment errors was provided to the correlation key's object (the `ErrorChannelSendingCorrelationKey` class). Then to implement `Future`s (assuming futures is what we want), you would then only need to extend the `ErrorChannelSendingCorrelationKey` to support your `CorrelationData` object.

What are your thoughts on this? Did I overlook something?

Reply:
I reworked a little.
I also removed the timeout in the `blockSend = true` use case.
Because if i am not connected to the broker, i really want to block until the messages was able to be sendet.
Or do you see any case where the `JCSMPStreamingPublishCorrelatingEventHandler` will not be alled?

But do you want to merge #63 first to let me rebase?

Reply:
> But do you want to merge #63 first to let me rebase?

Merged.

> Or do you see any case where the JCSMPStreamingPublishCorrelatingEventHandler will not be alled?

This should be a non-issue as long as the session is never explicitly closed. And we only ever close it on application shutdown, so I don't think this is something you need to be concerned about.